File: pnw_api.py
import requests
import time
import os

# Removed: API_URL = f"https://api.politicsandwar.com/graphql?api_key={API_KEY}" - URL will be built in run_query
RATE_LIMIT_DELAY = 0.1  # 1 second delay between requests

def run_query(api_key: str, query: str):
    """
    Run a GraphQL query against the Politics & War API.

    Args:
        api_key: The Politics & War API key.
        query: GraphQL query string

    Returns:
        JSON response data

    Raises:
        ValueError: If there is an API error, authentication error, or invalid response
    """
    # Check if API key is provided
    if not api_key:
        raise ValueError("API_KEY is not provided. Please enter your Politics & War API key.")

    API_URL = f"https://api.politicsandwar.com/graphql?api_key={api_key}"

    try:
        time.sleep(RATE_LIMIT_DELAY)  # Add delay between requests
        response = requests.post(API_URL, json={"query": query})

        # Handle specific HTTP error codes
        if response.status_code == 401:
            raise ValueError("API authentication failed. Check your API key.")
        elif response.status_code == 403:
            raise ValueError("API access forbidden. Your key may be invalid or lacks permissions.")
        elif response.status_code == 429:  # Too Many Requests
            print("Rate limit hit, waiting to retry...")
            time.sleep(5)  # Wait longer if we hit the rate limit
            response = requests.post(API_URL, json={"query": query})
            if response.status_code != 200:
                raise ValueError(f"Rate limit retry failed with status code {response.status_code}")
        elif response.status_code != 200:
            raise ValueError(f"API request failed with status code {response.status_code}")

        # Parse response as JSON
        data = response.json()

        # Check for GraphQL errors
        if "errors" in data:
            error_messages = [error.get("message", "Unknown GraphQL error") for error in data.get("errors", [])]
            error_message = "; ".join(error_messages)
            print(f"GraphQL API Error: {error_message}")
            raise ValueError(f"GraphQL API Error: {error_message}")

        # Validate response structure
        if "data" not in data:
            raise ValueError("API response missing 'data' field")

        return data

    except requests.exceptions.RequestException as e:
        # Handle network errors
        print(f"Network error communicating with the API: {str(e)}")
        raise ValueError(f"Network error: {str(e)}")
    except ValueError as e:
        # Re-raise ValueError for specific API errors
        raise
    except Exception as e:
        # Catch any other errors
        print(f"Unexpected error in API query: {str(e)}")
        raise ValueError(f"API query failed: {str(e)}")

def get_my_nation(api_key: str):
    """
    Get information about the currently authenticated nation.

    Args:
        api_key: The Politics & War API key.

    Returns:
        Nation data dictionary

    Raises:
        ValueError: If authentication fails or the API returns an error
    """
    query = """
    {
      me {
        nation {
          id
          nation_name
          score  
          soldiers
          tanks
          aircraft
          ships
          missiles
          nukes
          spies
          alliance {
            id
            name
            treaties {
              alliance1_id
              alliance2_id
              treaty_type
              treaty_url
            }
          }
        }
      }
    }
    """
    # Run the query with already enhanced error handling
    data = run_query(api_key, query)

    # Additional error handling for specific me/nation response errors
    if "data" not in data:
        raise ValueError("API response missing data field")
    if "me" not in data["data"]:
        raise ValueError("API response missing 'me' field - authentication may have failed")
    if not data["data"]["me"] or "nation" not in data["data"]["me"]:
        raise ValueError("Failed to retrieve nation data. Your API key may be invalid or expired.")

    return data["data"]["me"]["nation"]

def get_nation_by_id(api_key: str, nation_id: int):
    """
    Get information about a specific nation by its ID.

    Args:
        api_key: The Politics & War API key.
        nation_id: The ID of the nation to retrieve.

    Returns:
        Nation data dictionary for the specified nation.

    Raises:
        ValueError: If the nation is not found or the API returns an error.
    """
    query = f"""
    query {{
      nations(id: {nation_id}, first: 1) {{
        data {{
          id
          nation_name
          score
          soldiers
          tanks
          aircraft
          ships
          missiles
          nukes
          spies
          color
          vacation_mode_turns
          alliance_id
          gross_national_income
          cities {{
            supermarket
            bank
            shopping_mall
            stadium
            subway
          }}
          wars {{
            turns_left
            date
            def_id
            attacks {{
              def_id
              money_stolen
              date
            }}
          }}
          defensive_wars_count
          beige_turns
          alliance {{
            id
            name
            treaties {{
              alliance1_id
              alliance2_id
              treaty_type
              treaty_url
            }}
          }}
        }}
      }}
    }}
    """
    # Run the query - error handling for network and basic API errors happens in run_query
    data = run_query(api_key, query)

    # Validate response structure specific to this query
    if "data" not in data or not data["data"]:
        raise ValueError("API response missing 'data' field or 'data' is null.")
    
    if "nations" not in data["data"]:
        raise ValueError("API response missing 'nations' field.")
    
    if "data" not in data["data"]["nations"] or not isinstance(data["data"]["nations"]["data"], list):
        raise ValueError("API response missing 'nations.data' field or it's not a list.")

    if not data["data"]["nations"]["data"]:
        raise ValueError(f"Nation with ID {nation_id} not found.")

    return data["data"]["nations"]["data"][0]

def get_nations(api_key: str, page=1):
    """
    Get a list of nations from the Politics & War API.

    Args:
        api_key: The Politics & War API key.
        page: Page number for pagination

    Returns:
        Dictionary containing nation data and pagination info

    Raises:
        ValueError: If the API returns an error or unexpected response structure
    """
    query = """
    {{
      nations(page: {page}, first: 500) {{
        data {{
          id
          nation_name
          score
          alliance_id
          vacation_mode_turns
          beige_turns
          color
          soldiers
          tanks
          aircraft
          ships
          missiles
          nukes
          spies
          gross_national_income
          cities {{
            supermarket
            bank
            shopping_mall
            stadium
            subway
          }}
          alliance {{
            id
            name
            treaties {{
              alliance1_id
              alliance2_id
              treaty_type
            }}
          }}
          wars {{
            turns_left
            date
            def_id
            attacks {{
              def_id
              money_stolen
              date
            }}
          }}
          defensive_wars_count
        }}
        paginatorInfo {{
          hasMorePages
          currentPage
        }}
      }}
    }}
    """.format(page=page)
    # Run the query - error handling happens in run_query function
    data = run_query(api_key, query)

    # Additional validation for this specific endpoint
    if "data" not in data:
        raise ValueError("API response missing 'data' field")
    if "nations" not in data["data"]:
        raise ValueError("API response missing 'nations' field - API format may have changed")
    if "data" not in data["data"]["nations"]:
        raise ValueError("API response missing nation data")

    # Log success and nation count
    nation_count = len(data["data"]["nations"]["data"])
    print(f"Successfully fetched {nation_count} nations from API (page {page})")

    return data["data"]["nations"]
    """
    Get a list of beige nations with 1 beige turn left.

    Args:
        api_key: The Politics & War API key.
        args: An object containing filtering parameters (limit, max_pages, ignore_dnr).
        min_score_ratio: Minimum target score ratio relative to user's nation score.
        max_score_ratio: Maximum target score ratio relative to user's nation score.

    Returns:
        Tuple containing user's nation data and a list of beige nations.

    Raises:
        ValueError: If the API returns an error or unexpected response structure.
    """
    my_nation = get_my_nation(api_key)
    filtered_beige_nations = []
    page = 1
    has_more_pages = True

    while has_more_pages and page <= args.max_pages:
        nations_data = get_nations(api_key, page)
        
        # Filter nations on the current page
        filtered_on_page = filter_beige_targets(nations_data['data'], my_nation, args, min_score_ratio, max_score_ratio)
        
        # Add filtered nations from this page to the overall list
        filtered_beige_nations.extend(filtered_on_page)
        
        # Check if we have reached the limit
        if len(filtered_beige_nations) >= args.limit:
            print(f"Limit of {args.limit} reached, stopping page fetching.")
            break # Stop fetching pages if limit is reached
            
        has_more_pages = nations_data['paginatorInfo']['hasMorePages']
        page += 1

    # Sort and apply limit to the collected filtered nations
    # The filter_beige_targets function already sorts and limits, but we need to re-sort
    # and limit the combined list from multiple pages if the limit wasn't exactly met
    # when breaking the loop. However, since filter_beige_targets now filters page by page,
    # we need to sort and limit the accumulated list here.
    
    # Re-sort the accumulated filtered nations by least beige turns
    filtered_beige_nations.sort(key=lambda x: x.get('beige_turns', 0))

    # Apply the limit to the final sorted list
    final_filtered_nations = filtered_beige_nations[:args.limit]


    return my_nation, final_filtered_nations

def has_treaty(my_alliance, target_alliance, protected_types=None):
    """
    Check if two alliances have a treaty that should prevent raiding.

    Args:
        my_alliance: Alliance data for your nation
        target_alliance: Alliance data for target nation
        protected_types: List of treaty types that prevent raiding

    Returns:
        Boolean indicating if a treaty exists
    """
    if not protected_types:
        # Default treaty types that should prevent raiding
        protected_types = ['MDP', 'MDOAP', 'ODP', 'ODOAP', 'NAP', 'PIAT', 'Protectorate']

    if not my_alliance or not target_alliance:
        return False

    # Check both alliances' treaties
    for treaty in my_alliance.get('treaties', []):
        if (treaty['alliance1_id'] == target_alliance['id'] or
            treaty['alliance2_id'] == target_alliance['id']):
            if treaty['treaty_type'] in protected_types:
                return True

    return False

File: lib/pnw-api.ts
interface ApiResponse {
  data: any
}

interface Nation {
  id: number
  nation_name: string
  score: number
  soldiers: number
  tanks: number
  aircraft: number
  ships: number
  missiles: number
  nukes: number
  spies: number
  alliance?: {
    id: number
    name: string
    treaties?: Treaty[]
  }
  alliance_id?: number
  vacation_mode_turns?: number
  beige_turns?: number
  color?: string
  gross_national_income?: number
  cities?: City[]
  wars?: War[]
  defensive_wars_count?: number
}

interface City {
  supermarket: number
  bank: number
  shopping_mall: number
  stadium: number
  subway: number
  infrastructure?: number
}

interface War {
  turns_left: number
  date: string
  def_id: number
  attacks?: Attack[]
}

interface Attack {
  def_id: number
  money_stolen: number
  date: string
}

interface Treaty {
  alliance1_id: number
  alliance2_id: number
  treaty_type: string
  treaty_url?: string
}

const RATE_LIMIT_DELAY = 100 // 100ms delay between requests

export async function runQuery(apiKey: string, query: string): Promise<ApiResponse> {
  console.log("Running GraphQL query", { queryLength: query.length })
  
  if (!apiKey) {
    throw new Error("API_KEY is not provided. Please enter your Politics & War API key.")
  }

  const API_URL = `https://api.politicsandwar.com/graphql?api_key=${apiKey}`

  try {
    // Add delay between requests
    await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_DELAY))
    
    console.log("Making request to PnW API")
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ query }),
    })

    console.log("API response received", { status: response.status })

    // Handle specific HTTP error codes
    if (response.status === 401) {
      throw new Error("API authentication failed. Check your API key.")
    } else if (response.status === 403) {
      throw new Error("API access forbidden. Your key may be invalid or lacks permissions.")
    } else if (response.status === 429) {
      console.log("Rate limit hit, waiting to retry...")
      await new Promise(resolve => setTimeout(resolve, 5000))
      
      const retryResponse = await fetch(API_URL, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ query }),
      })
      
      if (retryResponse.status !== 200) {
        throw new Error(`Rate limit retry failed with status code ${retryResponse.status}`)
      }
      
      const retryData = await retryResponse.json()
      return retryData
    } else if (response.status !== 200) {
      throw new Error(`API request failed with status code ${response.status}`)
    }

    // Parse response as JSON
    const data = await response.json()
    console.log("API data parsed", { hasData: !!data.data, hasErrors: !!data.errors })

    // Check for GraphQL errors
    if (data.errors) {
      const errorMessages = data.errors.map((error: any) => error.message || "Unknown GraphQL error")
      const errorMessage = errorMessages.join("; ")
      console.error("GraphQL API Error:", errorMessage)
      throw new Error(`GraphQL API Error: ${errorMessage}`)
    }

    // Validate response structure
    if (!data.data) {
      throw new Error("API response missing 'data' field")
    }

    return data

  } catch (error) {
    console.error("API query error:", error)
    if (error instanceof Error) {
      throw error
    }
    throw new Error(`API query failed: ${error}`)
  }
}

export async function getNationById(apiKey: string, nationId: number): Promise<Nation> {
  console.log("Getting nation by ID", { nationId })
  
  const query = `
    query {
      nations(id: ${nationId}, first: 1) {
        data {
          id
          nation_name
          score
          soldiers
          tanks
          aircraft
          ships
          missiles
          nukes
          spies
          color
          vacation_mode_turns
          alliance_id
          gross_national_income
          cities {
            supermarket
            bank
            shopping_mall
            stadium
            subway
          }
          wars {
            turns_left
            date
            def_id
            attacks {
              def_id
              money_stolen
              date
            }
          }
          defensive_wars_count
          beige_turns
          alliance {
            id
            name
            treaties {
              alliance1_id
              alliance2_id
              treaty_type
              treaty_url
            }
          }
        }
      }
    }
  `
  
  const data = await runQuery(apiKey, query)
  
  // Validate response structure
  if (!data.data?.nations?.data || !Array.isArray(data.data.nations.data)) {
    throw new Error("API response missing 'nations.data' field or it's not a list.")
  }

  if (data.data.nations.data.length === 0) {
    throw new Error(`Nation with ID ${nationId} not found.`)
  }

  console.log("Nation data retrieved", { nationName: data.data.nations.data[0].nation_name })
  return data.data.nations.data[0]
}

export async function getNations(apiKey: string, page: number = 1): Promise<any> {
  console.log("Getting nations page", { page })
  
  const query = `
    {
      nations(page: ${page}, first: 500) {
        data {
          id
          nation_name
          score
          alliance_id
          vacation_mode_turns
          beige_turns
          color
          soldiers
          tanks
          aircraft
          ships
          missiles
          nukes
          spies
          gross_national_income
          cities {
            supermarket
            bank
            shopping_mall
            stadium
            subway
          }
          alliance {
            id
            name
            treaties {
              alliance1_id
              alliance2_id
              treaty_type
            }
          }
          wars {
            turns_left
            date
            def_id
            attacks {
              def_id
              money_stolen
              date
            }
          }
          defensive_wars_count
        }
        paginatorInfo {
          hasMorePages
          currentPage
        }
      }
    }
  `
  
  const data = await runQuery(apiKey, query)
  
  // Additional validation for this specific endpoint
  if (!data.data?.nations) {
    throw new Error("API response missing 'nations' field - API format may have changed")
  }
  
  if (!data.data.nations.data) {
    throw new Error("API response missing nation data")
  }

  const nationCount = data.data.nations.data.length
  console.log(`Successfully fetched ${nationCount} nations from API (page ${page})`)

  return data.data.nations
}

export function hasTreaty(myAlliance: any, targetAlliance: any, protectedTypes?: string[]): boolean {
  const defaultProtectedTypes = ['MDP', 'MDOAP', 'ODP', 'ODOAP', 'NAP', 'PIAT', 'Protectorate']
  const treatyTypes = protectedTypes || defaultProtectedTypes

  if (!myAlliance || !targetAlliance) {
    return false
  }

  // Check both alliances' treaties
  const treaties = myAlliance.treaties || []
  for (const treaty of treaties) {
    if ((treaty.alliance1_id === targetAlliance.id || treaty.alliance2_id === targetAlliance.id)) {
      if (treatyTypes.includes(treaty.treaty_type)) {
        return true
      }
    }
  }

  return false
}